nbodykit/plugins/datasource/Zheng07HOD.py

```diff
@@ -0,0 +1,177 @@
+from nbodykit.extensionpoints import DataSource, datasources
+
+import numpy
+import logging
+
+logger = logging.getLogger('Zheng07HOD')
+
+class Zheng07HODDataSource(DataSource):
+    """
+    A `DataSource` that uses the HOD prescription of 
+    Zheng et al. 2007 to populate an input halo catalog with galaxies, 
+    and returns the (Position, Velocity) of those galaxies
+    
+    The mock population is done using `halotools` (http://halotools.readthedocs.org)
+    The HOD model is of the commonly-used form:
+    
+    * logMmin: Minimum mass required for a halo to host a central galaxy
+    * sigma_logM: Rate of transition from <Ncen>=0 --> <Ncen>=1
+    * alpha: Power law slope of the relation between halo mass and <Nsat>
+    * logM0: Low-mass cutoff in <Nsat>
+    * logM1: Characteristic halo mass where <Nsat> begins to assume a power law form
+    
+    See the documentation for the `halotools` builtin Zheng07 HOD model, 
+    for further details regarding the HOD 
+    """
+    plugin_name = "Zheng07HOD"
+    
+    def __init__(self, halocat, redshift, logMmin=13.031, sigma_logM=0.38, 
+                    alpha=0.76, logM0=13.27, logM1=14.08, rsd=None):
+        """
+        Default values for HOD values from Reid et al. 2014
+        """
+        # load halotools
+        try:
+            from halotools import sim_manager
+        except:
+            name = self.__class__.__name__
+            raise ValueError("`halotools` must be installed to use '%s' DataSource" %name)
+            
+        # need cosmology
+        if self.cosmo is None:
+            raise AttributeError("a cosmology instance is required to populate an HOD")
+        
+        # set defaults and load the rest
+        sim_manager.sim_defaults.default_cosmology = self.cosmo.engine
+        sim_manager.sim_defaults.default_redshift = self.redshift
+        from halotools import empirical_models as em_models
+
+        # explicitly set an analytic mass-concentration relation
+        sats_prof_model = em_models.NFWPhaseSpace(conc_mass_model='dutton_maccio14')
+        
+        # build the model and set defaults
+        base_model = em_models.PrebuiltHodModelFactory('zheng07')
+        self.model = em_models.HodModelFactory(baseline_model_instance=base_model, 
+                                                satellites_profile=sats_prof_model)
+        for param in self.model.param_dict:
+            self.model.param_dict[param] = getattr(self, param)
+        
+        # data from the halo catalog
+        [[Position, Velocity, Mass]] = self.halocat.read(['Position','Velocity', 'Mass'], full=True)
```

Comment:
That's exactly the point -- nbodykit is parallel, but halotools is not, so we need to adapt between the two. What we have here is each rank calls halotools with its local partition of the halos, and obtain a local set of galaxies.  (scenario 1, current implementation)

The alternative is all halos are gathered to a single rank, then halotools is called, then, finally the produced galaxy catalogue is distributed to the ranks. (scenario 2, serialized implementation)

Are these two processes equivalent? I worry about correlation of random number generators, and the possibility of missing global constraints with scenario 1. I agree runtime is not a concern, but correctness is.